zip-0222.rst

```diff
@@ -0,0 +1,314 @@
+====================== ==================================== ===============
+
+
+Consensus rules
+---------------
+
+Once the TODO network upgrade activates, the following new consensus rules are enforced:
+
+- For each ``(outpoint, witness)`` pair in ``tze_inputs``:
+
+  - ``outpoint`` MUST reference a precondition of the same type and mode in an already-mined
+    transaction.
+  - ``tze_verify(mode, precondition, witness, context)`` MUST return ``true``.
+
+- If a transaction has non-empty ``tze_inputs`` and non-empty ``tze_outputs``, then every
+  element in both fields MUST have the same ``type``.
+
+- Non-coinbase transactions MUST have at least one of the following:
+  - nonempty transparent inputs
+  - nonempty shielded inputs
+  - nonempty ``tze_inputs``
+
+- Transactions MUST have at least one of the following:
+  - nonempty transparent outputs
+  - nonempty shielded outputs
+  - nonempty ``tze_outputs``
+
+- All ``amount`` field values of ``tze_output`` records MUST be nonnegative and less than
+  MAX_MONEY.
+
+- The sum of amounts associated with all transparent, shielded, and TZE outputs within 
+  a transaction MUST NOT exceed the sum of the amounts associated with transaction inputs.
+
+Changes to signatures over transaction digests
+----------------------------------------------
+
+The newly added parts of the transaction will be included in transaction 
+digests for signatures. See [#zip-0244]_ for the specification of the new
+digests.
+
+Rationale
+=========
+
+Transactions that have both TZE inputs and outputs are required to use a single extension
+type, in order to prevent cross-protocol attacks. The downside is that this prevents all
+TZE-encumbered value from being spent directly into a different TZE type; the value needs
+to go through a regular address in between. This restriction might be relaxed in future
+ZIPs for specific combinations of ``(type, mode)`` pairs that have been analyzed for
+cross-protocol attacks, but we opt here for a fail-safe default behaviour.
+
+- TODO: Maybe also disallow regular (transparent, Sprout, and Sapling) spends/outputs in
+  transactions that have TZE inputs and outputs?
+
+An earlier draft version of this ZIP stored the payloads inside transparent inputs and
+outputs. Although this had the advantage of not requiring a transaction format change,
+the consensus rules were significantly more complicated, and the design coupled the
+extension logic too tightly to the transparent address logic. Instead, this ZIP uses
+dedicated transaction fields.
+
+
+Security and Privacy Considerations
+===================================
+
+This ZIP assumes that the base transaction format is non-malleable. However, the
+``precondition`` and ``witness`` byte sequences are treated here as opaque. It is the
+responsibility of ``tze_verify`` to enforce the following:
+
+- ``witness`` MUST be non-malleable: any malleation MUST cause ``tze_verify`` to
+  return ``false``.
```

Comment:
```suggestion
- ``precondition`` MUST be non-malleable: any malleation MUST cause ``tze_verify`` to
  return ``false``.
```
This would likely be implemented by extensions using a transaction digest internally as part of their protocols.

`witness` doesn't need to be non-malleable any longer; we should note here why (that it will be treated as "authorizing data", and not part of the standard transaction ID).